doc/news/_preparation_next_release.md

```diff
@@ -102,7 +102,14 @@ Known limitations:
 
 For more information see [Spec Plugin](../../src/plugins/spec/README.md).
 
-### <<HIGHLIGHT>>
+### Go binding into libelektra repository
+
+The [go binding](../../src/bindings/go-elektra/README.md) is now in the bindings folder.
+In your source code you can access the package with `github.com/ElektraInitiative/libelektra/go-elektra`.
```

Comment:
Ok, let’s do it this way. New url needs to be https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra.

Reply:
We need to adapt `go-import` metakey here, so `go get` works correctly. Instead of `https://github.com/ElektraInitiative/libelektra/go-elektra` we need to use ` https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra`.

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="go-import"
      content="go.libelektra.org git https://github.com/ElektraInitiative/libelektra/go-elektra"
    />
  </head>
  <body></body>
</html>
```

Reply:
As I assumed, this could cause problems. `go.libelektra.org` is already published to go packages and there is no way to remove it. I would suggest the other url I wrote in a comment above, `goelektra.libelektra.org`.

Reply:
Here is the PR https://github.com/ElektraInitiative/libelektra/pull/4929 which changes the url.

Reply:
Since we already changed `go.libelektra.org` and I assume this release will be completed ASAP (hours not days), I'd say keep things as they are right now. That means external code is currently broken, because `go.libelektra.org` points nowhere, but hopefully the release is done soon and then things should work again.


What we should have done in the first place:
1. Keep `go.libelektra.org` pointed at the old repo for now and use the direct `github.com/ElektraInitiative` URLs in our Go code (exactly as it was when #4895 was merged).
2. Only once the release is fully published, do we update `go.libelektra.org` to the new URL.

That way the go module would never be broken, it would just take a little longer until you get the new package with `go.libelektra.org`.

Reply:
Ok, would you suggest changing the url in https://github.com/ElektraInitiative/libelektra/pull/4929  to go.libelektra.org and after the release was done, merging it? Do I understand correctly?

Reply:
Yes, the URL used in our Go code in the repo should be changed after the release. For now keep the github.com URLs because those definitely work.

Reply:
Alright!

Reply:
The other PR is now prepared. This release should also be ready after pipeline is green.

Reply:
Do we need to do another Release (0.10.1) after #4929 is merged?

Reply:
Depends on when the version tag is created. @kodebach is it during the release process automatically?

Reply:
Would be good if we could manage to have our next version as `v0.10.0`

Reply:
> Do we need to do another Release (0.10.1) after #4929 is merged?

Don't think so. The PR will just change our internal code. It'll just get shipped with the next release.

Reply:
As noted in https://github.com/ElektraInitiative/libelektra/pull/4930#discussion_r1195182589 I don't actually think we need any extra step before or after the release (I'm made some incorrect assumptions). We just need to set up `go.libelektra.org` correctly.

However, AFAICT the current setup for `go.libelektra.org` is incorrect. With a correct setup it should be possible to download the new module with a commit hash even before the release with:

```sh
go mod download go.libelektra.org@<commit-hash>
```

However, if we try this (with `-x` to see what's happening), we get:

```sh
$ go mod download -x go.libelektra.org@567bb7fe5a89de0812da3d9452f61bcf70369fa3
# get https://proxy.golang.org/go.libelektra.org/@v/567bb7fe5a89de0812da3d9452f61bcf70369fa3.info
# get https://proxy.golang.org/go.libelektra.org/@v/567bb7fe5a89de0812da3d9452f61bcf70369fa3.info: 404 Not Found (4.455s)
# get https://go.libelektra.org/?go-get=1
# get https://go.libelektra.org/?go-get=1: 200 OK (0.159s)
mkdir -p /home/klemens/go/pkg/mod/cache/vcs # git3 https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra
# lock /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808.lock
# /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808 for git3 https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra
cd /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808; git -c log.showsignature=false log --no-decorate -n1 '--format=format:%H %ct %D' 567bb7fe5a89de0812da3d9452f61bcf70369fa3 --
0.007s # cd /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808; git -c log.showsignature=false log --no-decorate -n1 '--format=format:%H %ct %D' 567bb7fe5a89de0812da3d9452f61bcf70369fa3 --
cd /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808; git ls-remote -q origin
0.322s # cd /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808; git ls-remote -q origin
# get https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra
# get https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra: 200 OK (0.409s)
cd /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808; git tag -l
0.001s # cd /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808; git tag -l
go: go.libelektra.org@567bb7fe5a89de0812da3d9452f61bcf70369fa3: invalid version: git ls-remote -q origin in /home/klemens/go/pkg/mod/cache/vcs/abd419aa0842dc8b3a10fae37508bbfb31205b0da724977f07d41ebec527e808: exit status 128:
	fatal: repository 'https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra/' not found
```

As you can see `go` tries to clone the git repo as defined by `go.libelektra.org`, but that obviously doesn't work, because `https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/go-elektra` is not a git repo.

Furthermore, the go module system seems really stupid and you pretty much incompatible with monorepos (which is extra weird, since Google is a big proponent of monorepos). The best solution for having a go module at a custom path in the repo (without having to use e.g. `go.libelektra.org/src/bindings/go-elektra/kdb`) is to set up a custom Go module proxy and make users add it to their `GOPROXY` env-var or have a dedicated repo for the go-module which gets automated commits with every release (https://stackoverflow.com/a/72370919).

Reply:
Yes, I have found this today as well. We would also need to change `go-import` metatag to use `https://github.com/ElektraInitiative/libelektra``.

**Suggested:**

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="go-import"
      content="go.libelektra.org git https://github.com/ElektraInitiative/libelektra"
    />
  </head>
  <body></body>
</html>
```

Reply:
> Furthermore, the go module system seems really stupid and you pretty much incompatible with monorepos (which is extra weird, since Google is a big proponent of monorepos). The best solution for having a go module at a custom path in the repo (without having to use e.g. `go.libelektra.org/src/bindings/go-elektra/kdb`) is to set up a custom Go module proxy and make users add it to their `GOPROXY` env-var or have a dedicated repo for the go-module which gets automated commits with every release

By the second option, you mean still having the old `go-elektra` (now archived) repo and when we create releases in this repo here, we have an automated scripts that creates the same tag in the other repo and pushes it. Or am I wrong?
If yes, we again don‘t include the go binding into our build process, which was the initial „problem“. What do you think @markus2330 ? Is such a proxy server solution feasible?

Reply:
I think something like go.libelektra.org/src/bindings/go-elektra/kdb is fine, too. Requiring a GOPROXY sounds like extra-effort for users.

> If yes, we again don‘t include the go binding into our build process, which was the initial „problem“. What do you think @markus2330 ?

The goals are:

- that everything is included in our build process
- users have a somewhat same URL to use Elektra's go binding

Reply:
And to be fair, it is fine if external programs use `go.libelektra.org/src/bindings/go-elektra/kdb`, then we don‘t need to do any of those things, would you agree having such url here? @markus2330 

and yes, we also need to adapt the `go-import` meta tag as described above.

Reply:
> I think something like go.libelektra.org/src/bindings/go-elektra/kdb is fine, too. Requiring a GOPROXY sounds like extra-effort for users.
> 
> > If yes, we again don‘t include the go binding into our build process, which was the initial „problem“. What do you think @markus2330 ?
> 
> The goals are:
> 
> * that everything is included in our build process
> * users have a somewhat same URL to use Elektra's go binding

Perfect, then we just need to adapt the `go-import` to:
```html

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="go-import"
      content="go.libelektra.org git https://github.com/ElektraInitiative/libelektra"
    />
  </head>
  <body></body>
</html>

```

Reply:
done

Reply:
> By the second option, you mean still having the old go-elektra (now archived) repo

Kind of, yes. However, everything would still be developed, built and tested in the `libelektra` repo and everybody could use this version directly via `github.com/ElektraInitiative/libelektra/src/bindings/go-elektra/kdb` instead of `go.libelektra.org/kdb`.

To make `go.libelektra.org/kdb` work too (for release versions), we would have a new step in the release process. This step would create a file copy of `src/bindings/go-elektra` and commit it as a new commit to a separate `go-elektra` repo. This commit is also tagged with the release version. The `meta` tag on `go.libelektra.org` would point at this separate repo (Sidenote: `https://go.libelektra.org/kdb` may also need to have the HTML file). 

This would be the most seamless for users, since the could continue using `go.libelektra.org/kdb` for imports.

> I think something like go.libelektra.org/src/bindings/go-elektra/kdb is fine, too. Requiring a GOPROXY sounds like extra-effort for users.

`GOPROXY` is extra effort for _all_ users, yes, and also quite a bit of effort for us. I don't like that option either.

`go.libelektra.org/src/bindings/go-elektra/kdb` is definitely the easiest solution for us, but it's a pretty ugly path and the path would also change, if we were to move the go binding within the repo at some point. That is something I would normally not consider a breaking change.

---

To illustrate the 3 "good" options we have, I'll compare them with the Java binding:

- **Java**:
  - Users depend on the binding by adding a Maven/Gradle dependency on the artifact `org.libelektra:libelektra` (name defined in Gradle).
  - The artifact is created via a release process that uploads it to Maven Central.
  - In the repo the files live in `src/bindings/jna/libelektra/src/main/java/org/libelektra`. This path consists of four parts `<binding-root>/<project-root>/<source-root>/<namespace>`:
    - `<binding-root>`=`src/bindings/jna`
    - `<project-root>`=`libelektra`
    - `<source-root>`=`src/main/java`
    - `<namespace>`=`org/libelektra`
    
    Only changes to `<namespace>` part would be breaking changes that affect users of the binding. Changing `<project-root>` or `<source-root>` would only need changes in our internal Gradle files and changing `<binding-root>` would only need changes in CMake (if it requires any changes at all).
- **Go** with full path:
  - Users depend on the binding by adding an `import` with URL `go.libelektra.org/src/bindings/go-elektra`.
  - There is no additional release process, only a tag in the main libelektra repo is needed.
  - The files live in `src/bindings/go-elektra/kdb`. This consists of two parts `<binding-root>/<module-root>`.
    - `<binding-root>`=`src/bindings/go-elektra`
    - `<module-root>`=`kdb`
  
    Changes to both `<module-root>` are breaking, since this is equivalent to changing the namespace in Java. However, changing `<binding-root>` is also breaking, because users have to specify it explicitly.
    
- **Go** with `go-elektra` release repo:
  - Users depend on the binding by adding an `import` with URL `go.libelektra.org`.
  - During the release process the binding files are copied to a separate `go-elektra` repo (in principle similar to uploading to Maven Central)
  - The files live in `src/bindings/go-elektra/kdb`. This consists of two parts `<binding-root>/<module-root>`.
    - `<binding-root>`=`src/bindings/go-elektra`
    - `<module-root>`=`kdb`

     Similar to Java `<module-root>` is breaking, but `<binding-root>` can be changed without limitation as long as we update our release process accordingly.

Reply:
> `go.libelektra.org/src/bindings/go-elektra/kdb` is definitely the easiest solution for us, but it's a pretty ugly path and the path would also change, if we were to move the go binding within the repo at some point. That is something I would normally not consider a breaking change.

But this is a general problem in go. If you have a separate repo for any go module and it changes somehow, users will also be required to change, so I don't think we gain much by investing a lot of time into this. Therefore I will stick with having  `go.libelektra.org/src/bindings/go-elektra/kdb`.